src/dftbp/dftb/elecconstraints.F90

```diff
@@ -7,428 +7,396 @@
-    speciesL(orb%posShell(conSh,iSp):orb%posShell(conSh+1,iSp)-1,orb%posShell(conSh,iSp): &
-        & orb%posShell(conSh+1,iSp)-1,2) = -real(Lplus(1:2*iSh+1,1:2*iSh+1))
-    speciesL(orb%posShell(conSh,iSp):orb%posShell(conSh+1,iSp)-1,orb%posShell(conSh,iSp): &
-        & orb%posShell(conSh+1,iSp)-1,3) = aimag(Lz(1:2*iSh+1,1:2*iSh+1))
+    !> Total shift of all constraints
+    real(dp), intent(out) :: shift(:,:,:,:)
 
-    allocate(tmpBlock(orb%mOrb,orb%mOrb))
+    !! Iterates over constraints
+    integer :: iConstr
 
-    Lshell = 0.0_dp
+    !! Number of constraints requested by the user
+    integer :: nConstr
 
-    iOrb = orb%nOrbSpecies(iSp)
-    tmpBlock(:,:) = 0.0_dp
+    shift(:,:,:,:) = 0.0_dp
+    nConstr = size(this%wAt)
 
-    ! identity part
-    tmpBlock(1:iOrb,1:iOrb) = qBlockSkew(1:iOrb,1:iOrb,conAt,1)
-    iStart = orb%posShell(conSh,iSp)
-    iEnd = orb%posShell(conSh+1,iSp)-1
-    do ii = 1, 3
-      Lshell(ii) = &
-          & - sum(SpeciesL(iStart:iEnd,iStart:iEnd,ii) &
-          &  * transpose(tmpBlock(iStart:iEnd,iStart:iEnd)))
+    do iConstr = 1, nConstr
+      call getConstrainShiftQ(shift, this%Vc(iConstr), this%wAt(iConstr)%data,&
+          & this%wAtOrb(iConstr)%data, this%wAtSpin(iConstr)%data)
     end do
 
-    Sshell = sum(qIn(orb%posShell(conSh,species(conAt)): &
-        & orb%posShell(conSh+1,species(conAt))-1,conAt,2:4),dim=1)
+  end subroutine getConstrainShift
 
-    ! Push Mj towards required value
-    w = V * 0.5_dp*(dot_product(lshell,vecNorm)+ &
-        & 0.5_dp*dot_product(Sshell,vecNorm)-MjTarget)
 
-    do ii = 1, 3
-      iShift(iStart:iEnd,iStart:iEnd,conAt,1) = &
-          & iShift(iStart:iEnd,iStart:iEnd,conAt,1) &
-          & + w * vecNorm(ii) * SpeciesL(iStart:iEnd,iStart:iEnd,ii)
-    end do
+  !> Get shift for atomic charge constraint.
+  subroutine getConstrainShiftQ(shift, Vc, wAt, wOrb, wSp)
+
+    !> Shift to which contribution is appended
+    real(dp), intent(inout) :: shift(:,:,:,:)
+
+    !> Potential / Lagrange multiplier
+    real(dp), intent(in) :: Vc
+
+    !> Atom(s) involved in current constrain
+    integer, intent(in) :: wAt(:)
+
+    !> Orbital(s) involved in current constrain
+    integer, intent(in) :: wOrb(:)
+
+    !> Spin(s) involved in current constrain
+    real(dp), intent(in) :: wSp(:,:)
+
+    !! Number of spin channels to be constrained
+    integer :: nSpin
+
+    !! Index of spin channel
+    integer :: iSpin
+
+    !! Index of atomic orbital
+    integer :: iW
+
+    nSpin = size(wSp, dim=2)
 
-    do iSpin = 2, nSpin
-      do iOrb = orb%posShell(conSh,species(conAt)), &
-          & orb%posShell(conSh+1,species(conAt))-1
-        shift(iOrb,iOrb,conAt,iSpin) = shift(iOrb,iOrb,conAt,iSpin) &
-            & + w * vecNorm(iSpin-1)
+    do iSpin = 1, nSpin
+      do iW = 1, size(wAt)
+        shift(wOrb(iW), wOrb(iW), wAt(iW), iSpin) = shift(wOrb(iW), wOrb(iW), wAt(iW), iSpin)&
+            & + Vc * wSp(iW, iSpin)
```

Comment:
Don't we miss the overlap here?

Reply:
I will mark this as resolved, since the error seems to be located in the paper, not the implementation.